File: android/app-v2/src/main/java/com/smarthome/livingroom_v2/brain/MockBrainClient.kt
package com.smarthome.livingroom_v2.brain

import com.smarthome.livingroom_v2.brain.dto.EdgeNodeInfo
import com.smarthome.livingroom_v2.brain.dto.EdgeRegisterRequest
import com.smarthome.livingroom_v2.brain.dto.EdgeRegisterResponse
import com.smarthome.livingroom_v2.brain.dto.EdgeRegistration
import com.smarthome.livingroom_v2.brain.dto.ExecutionReport
import com.smarthome.livingroom_v2.brain.dto.Plan
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Local stand-in used only when remote Brain is unavailable (register/heartbeat/report).
 * Intent / plan pull is always from the real server via [HttpCommandSource] — no local plan queue.
 */
class MockBrainClient : BrainClient {
    private val registrations = ConcurrentHashMap<String, EdgeNodeInfo>()
    private val reports = CopyOnWriteArrayList<ExecutionReport>()
    private val issuedEdgeIds = ConcurrentHashMap.newKeySet<String>()

    val lastNodeInfo: EdgeNodeInfo?
        get() = registrations.values.maxByOrNull { it.reportedAtSec }

    val lastRegistration: EdgeRegistration?
        get() {
            val info = lastNodeInfo ?: return null
            return EdgeRegistration(
                edgeId = info.edgeId,
                services = info.services,
                registeredAt = (info.reportedAtSec * 1000).toLong(),
            )
        }

    fun snapshotReports(): List<ExecutionReport> = reports.toList()

    override suspend fun registerEdge(request: EdgeRegisterRequest): EdgeRegisterResponse {
        val hint = request.clientHint?.trim().orEmpty()
        val edgeId = when {
            hint.isNotEmpty() && hint !in issuedEdgeIds -> hint
            hint.isNotEmpty() -> "$hint-${UUID.randomUUID().toString().take(6).lowercase()}"
            else -> "edge-${UUID.randomUUID().toString().take(8).lowercase()}"
        }
        issuedEdgeIds.add(edgeId)
        return EdgeRegisterResponse(
            ok = true,
            status = "approved",
            edgeId = edgeId,
            message = "mock approved",
            ts = System.currentTimeMillis() / 1000.0,
        )
    }

    override suspend fun reportEdgeInfo(info: EdgeNodeInfo): Boolean {
        if (info.edgeId.isBlank()) {
            throw HttpEdgeException(401, "heartbeat requires edgeId; call registerEdge first")
        }
        issuedEdgeIds.add(info.edgeId)
        registrations[info.edgeId] = info
        return true
    }

    override suspend fun fetchPlans(edgeId: String): List<Plan> = emptyList()

    override suspend fun report(report: ExecutionReport) {
        reports.add(report)
        if (reports.size > 100) {
            reports.subList(0, reports.size - 100).clear()
        }
    }
}

File: ios/LivingRoomEdge/LivingRoomEdge/Brain/MockBrainClient.swift
import Foundation

/// In-process stand-in for the AI brain (mirrors Android MockBrainClient).
/// Isolated on the main actor — no NSLock (Swift 6 forbids lock/unlock in async contexts).
@MainActor
final class MockBrainClient: BrainClient, ObservableObject {
    private var registrations: [String: EdgeNodeInfo] = [:]
    private var pendingPlans: [String: [Plan]] = [:]
    private(set) var reports: [ExecutionReport] = []
    /// edgeIds issued by mock registerEdge.
    private var issuedEdgeIds: Set<String> = []

    var lastNodeInfo: EdgeNodeInfo? {
        registrations.values.max(by: { $0.reportedAt < $1.reportedAt })
    }

    /// Legacy alias.
    var lastRegistration: EdgeRegistration? {
        guard let info = lastNodeInfo else { return nil }
        return EdgeRegistration(edgeId: info.edgeId, services: info.services, registeredAt: info.reportedAt)
    }

    func nodeInfo(for edgeId: String) -> EdgeNodeInfo? {
        registrations[edgeId]
    }

    func enqueuePlan(edgeId: String, plan: Plan) {
        pendingPlans[edgeId, default: []].append(plan)
    }

    @discardableResult
    func enqueueSkillAction(
        edgeId: String,
        skillId: String,
        action: String,
        params: [String: String] = [:]
    ) -> Plan {
        let plan = Plan(
            planId: "plan-\(UUID().uuidString.prefix(8))",
            steps: [
                PlanStep(
                    stepId: "step-1",
                    skillId: skillId,
                    action: action,
                    params: params,
                    onFailure: .abort
                ),
            ]
        )
        enqueuePlan(edgeId: edgeId, plan: plan)
        return plan
    }

    func registerEdge(_ request: EdgeRegisterRequest) async throws -> EdgeRegisterResponse {
        let hint = (request.clientHint ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        let edgeId: String
        if !hint.isEmpty, !issuedEdgeIds.contains(hint) {
            edgeId = hint
        } else if !hint.isEmpty {
            edgeId = "\(hint)-\(String(UUID().uuidString.prefix(6)).lowercased())"
        } else {
            edgeId = "edge-\(String(UUID().uuidString.prefix(8)).lowercased())"
        }
        issuedEdgeIds.insert(edgeId)
        return EdgeRegisterResponse(
            ok: true,
            status: "approved",
            edgeId: edgeId,
            message: "mock approved"
        )
    }

    func reportEdgeInfo(_ info: EdgeNodeInfo) async throws {
        if info.edgeId.isEmpty {
            throw NSError(
                domain: "MockBrainClient",
                code: 401,
                userInfo: [NSLocalizedDescriptionKey: "heartbeat requires edgeId; call registerEdge first"]
            )
        }
        issuedEdgeIds.insert(info.edgeId)
        registrations[info.edgeId] = info
    }

    func fetchPlans(edgeId: String) async throws -> [Plan] {
        let drained = pendingPlans[edgeId] ?? []
        pendingPlans[edgeId] = []
        return drained
    }

    func report(_ report: ExecutionReport) async throws {
        reports.append(report)
        if reports.count > 100 {
            reports.removeFirst(reports.count - 100)
        }
    }
}
